src/main/java/de/tum/in/www1/artemis/service/hestia/TestwiseCoverageService.java

```diff
@@ -0,0 +1,233 @@
+     * count the same covered lines multiple times.
+     * @param coverageReport the report for which the line counts of its file reports should be caluclated and saved
+     * @return the number of covered lines by file path
+     */
+    private Map<String, Integer> calculateAndSaveUniqueLineCountsByFilePath(CoverageReport coverageReport) {
+        var report = coverageReportRepository.findCoverageReportByIdWithEagerFileReportsAndEntriesElseThrow(coverageReport.getId());
+        var coveredLinesByFilePath = new HashMap<String, Integer>();
+        report.getFileReports().forEach(fileReport -> {
+            var lineSet = new HashSet<Integer>();
+            fileReport.getTestwiseCoverageEntries()
+                    .forEach(entry -> lineSet.addAll(IntStream.rangeClosed(entry.getStartLine(), entry.getStartLine() + entry.getLineCount() - 1).boxed().toList()));
+            fileReport.setCoveredLineCount(lineSet.size());
+            coverageFileReportRepository.save(fileReport);
+            coveredLinesByFilePath.put(fileReport.getFilePath(), lineSet.size());
+        });
+        return coveredLinesByFilePath;
+    }
+
+    /**
+     * Returns the testwise coverage report for the latest solution submission for a programming exercise without the file reports
+     * @param programmingExercise the exercise for which the latest coverage report should be retrieved
+     * @return the testwise coverage report for the latest solution submission without the file reports
+     */
+    public CoverageReport getCoverageReportForLatestSolutionSubmissionFromProgrammingExercise(ProgrammingExercise programmingExercise) {
+        var solutionParticipation = solutionProgrammingExerciseParticipationRepository.findByProgrammingExerciseIdElseThrow(programmingExercise.getId());
+        var latestSubmissions = programmingSubmissionRepository.findLatestLegalSubmissionForParticipation(solutionParticipation.getId(), Pageable.ofSize(1));
+        var latestSubmission = latestSubmissions.get(0);
+        return coverageReportRepository.findCoverageReportBySubmissionIdElseThrow(latestSubmission.getId());
+    }
+
+    /**
+     * Returns the full testwise coverage report for the latest solution submission for a programming exercise containing all file reports
+     * @param programmingExercise the exercise for which the latest coverage report should be retrieved
+     * @return the full testwise coverage report for the latest solution submission with all file reports
+     */
+    public CoverageReport getFullCoverageReportForLatestSolutionSubmissionFromProgrammingExercise(ProgrammingExercise programmingExercise) {
+        var lazyReport = getCoverageReportForLatestSolutionSubmissionFromProgrammingExercise(programmingExercise);
+        return coverageReportRepository.findCoverageReportByIdWithEagerFileReportsAndEntriesElseThrow(lazyReport.getId());
```

Comment:
The query for loading the coverage report for the latest submission for a given programming exercise id:
```
SELECT DISTINCT r FROM CoverageReport r
LEFT JOIN FETCH r.submission s
WHERE s.participation.exercise.id = :#{#programmingExerciseId}
AND (s.type <> 'ILLEGAL' or s.type is null)
order by s.submissionDate desc
```

In terms of readability, do you think that another Query with fetching the file reports and entries should be implemented? Aside from the `double coveredLineRatio` of CoverageReport, no attributes/objects are loaded twice with both calls.

Reply:
Maybe my comment was confusing.
I understand that your suggestion is not about increasing the readability, I mean realizing it would decrease the readability (and maintainability). In that case, we would need two different repository methods:
1. The above one with 7 LOC.
2. The above one with "LEFT JOIN FETCH r.fileReports f" and "LEFT JOIN FETCH f.testwiseCoverageEntries". (9 LOC, 7 duplicated).

The difference between both calls is that the first call retrieves the CoverageReport for a ProgrammingExercise (id) while the second one loads the eager values for a given CoverageReport-ID.

Reply:
If you want you can leave it like this then.